### ISSUE TITLE
Improve reasoning for size oblivious equations involving min or max()

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#125914]
### 🐛 Describe the bug

Internal xref: https://fb.workplace.com/groups/6829516587176185/posts/7172183699576137/

This program fails to compile:

```
@torch.compile(fullgraph=True, backend="eager")
def cf(x):
    u0, u1 = x.tolist()
    torch._check_is_size(u0)
    torch._check_is_size(u1)
    torch._check(u0 + u1 == 20)
    if guard_size_oblivious(torch.sym_max(1, u0 + u1) == 20):
        return torch.tensor(True)
    else:
        return torch.tensor(False)

@run_test
def test_symmax():
    assert cf(torch.tensor([10, 10])).item()
```

Actually, we should be able to make the inference here, because u0 and u1 are size-like in a size oblivious, so we assume they are >= 2, which means that the Max should evaporate, but we are unable to do this.

cc @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @bobrenjc93 @bdhirsh @anijain2305 @lezcano 

### Versions

main

**Comment by ezyang** [comment]
cc @shazqadeer

**Comment by ezyang** [comment]
The harder version of this is when 20 is replaced with a (typically backed) symbolic variable s0.

**Comment by lezcano** [comment]
I would argue that, since you are literally checking u0+u1 == 20 in the line before, we should be able to pattern match this and use it. We should pattern match something that is expression == constant at the very least.

**Comment by shazqadeer** [comment]
I am surprised this does not work because we have a value range analysis in symbolic shapes. So the obvious rules for + and max should do the job within guard_size_oblivious (using the knowledge that u0 >= 2 and u1 >= 2).  Perhaps there is a bug in value range analysis for either + or max.

**Comment by shazqadeer** [comment]
> The harder version of this is when 20 is replaced with a (typically backed) symbolic variable s0.

Indeed. In this case, even the perfect value range analysis I talk about in my first comment will not address solve the problem. It seems we need some kind of equality propagation that interacts with value-range analysis appropriately. As far as I know, we don't have form of equality propagation implemented in symbolic shapes right now. But I suppose something simple could be done based on e-graphs.

**Comment by shazqadeer** [comment]
> I am surprised this does not work because we have a value range analysis in symbolic shapes. So the obvious rules for + and max should do the job within guard_size_oblivious (using the knowledge that u0 >= 2 and u1 >= 2). Perhaps there is a bug in value range analysis for either + or max.

```
a = ValueRanges(lower=1, upper=1, is_bool=False)              
b = ValueRanges(lower=0, upper=40, is_bool=False)
value range = ValueRanges(lower=1, upper=40, is_bool=False)
a = ValueRanges(lower=1, upper=1, is_bool=False)
b = ValueRanges(lower=0, upper=40, is_bool=False)            
value range = ValueRanges(lower=1, upper=40, is_bool=False)
a = ValueRanges(lower=1, upper=1, is_bool=False)                                                                                             
b = ValueRanges(lower=0, upper=40, is_bool=False)
value range = ValueRanges(lower=1, upper=40, is_bool=False)
```

I printed out the input and output value ranges in ```max```. The lower bounds appear to be incorrect. It looks like the value range analysis is not respecting ```guard_size_oblivious``` semantics.

**Comment by lezcano** [comment]
Value ranges cannot prove this, because you cannot refine the value range from the `u0 + u1 == 20` expression. 
The way to do this is by registering a replacement of `u0 + u1 -> 20` and adding this to our pattern matcher. But we currently do not have replacements of Expressions into numbers, we just have Symbols into numbers.
We should be able to implement this when all the symbols are on one side. It should be alright when one side is just unbacked symbols and the other is sizes. In the general case it gets trickier, figuring out which direction you can / should perform the replacement.

Ed and I discussed a crazier idea to solve this, which would be to enhance value ranges with support for expressions. Then when traversing down the expression tree, you'd try to do a pattern match into existing expressions, and stop the traversal and return the associated value range for a given subtree if there's a match. This would be more general in that it supports inequalities, but it's  notably more difficult to implement.

**Comment by shazqadeer** [comment]
> Value ranges cannot prove this, because you cannot refine the value range from the u0 + u1 == 20 expression.

Understood. And I agree.

**Comment by ColinPeppler** [comment]
Seeing this come up more often.

`Eq(-Min(1, u7) + Min(9223372036854775807, u7), -Min(0, u7) + Min(u7, Max(0, u7 - 1))`
e.g. from broadcast_symbolic_shapes
https://github.com/pytorch/pytorch/blob/938803df945c83b964fd9f3e4f55494b0a940806/torch/_inductor/lowering.py#L452


`Max(1, u1) <= Max(1, u1)*Max(1, u1)` 
e.g. from is_stride_ordered
https://github.com/pytorch/pytorch/blob/53c9c19e762576ea86ab0666c1b1e070b1993857/torch/_inductor/ir.py#L3035

**Comment by ezyang** [comment]
Another one from @jbschlosser `LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(-Min(22, Max(0, u0)) + Min(22, Max(u0 + u1, Max(0, u0))), 0) (unhinted: Eq(-Min(s0, Max(0, u0)) + Min(s0, Max(u0 + u1, Max(0, u0))), 0)).  (Size-like symbols: u1, u0)`

**Comment by jbschlosser** [comment]
Context for my issue is NJT `narrow(..., dim=0)` -> dense `narrow(..., dim=0)` -> dense `slice()` underneath. For my example:
* `u0`: `start` passed to `narrow()`
* `u1`: `length` passed to `narrow()`
* `(22, 5)` is the shape of the dense tensor on which `narrow()` is operating

The clamping logic here contributes to the error:
https://github.com/pytorch/pytorch/blob/41e59754b407533b060b874c22ca4feda38bd83a/torch/_inductor/ir.py#L3014

**Comment by yushangdi** [comment]
Can still repro on master.

@lezcano @ezyang  I'm helping with "old issue scrubbing" this quarter. Just checking in here. Is this task still relevant and still on your plate?

**Comment by bobrenjc93** [comment]
@yushangdi can you share your repro? https://github.com/pytorch/pytorch/pull/147344/files should have fixed this and added the repro as a test.

**Comment by yushangdi** [comment]
> [@yushangdi](https://github.com/yushangdi) can you share your repro? https://github.com/pytorch/pytorch/pull/147344/files should have fixed this and added the repro as a test.

oh oops, sorry, I meant to comment on this one instead https://github.com/pytorch/pytorch/issues/132522
